examples/quantum_variation.py

```diff
@@ -0,0 +1,327 @@
+            c.x(0)
+            U_gate(i, c)
+            c.x(0)
+        if i == q:
+            U_gate(i, c)
+            R_gate(i, c)
+            break
+        R_gate(i, c)
+    pstar = np.real(np.array(c.expectation([np.array([[1, 1], [1, 1]]) / 2, [0]])))
+    return mod * (2 * pstar - 1)
+
+
+def Calculation_C(mod, theta_x, k, q):
+    # mod: a in paper; theta_x: theta in paper; k, q: A[k, q] or C[k] qth term; whi: whi=0 A whi=1 C
+    ancilla = np.array([1, np.exp(1j * theta_x)]) / np.sqrt(2)
+    c = tc.Circuit(N + 1, inputs=np.kron(ancilla, state))
+    for i in range(len(door)):
+        if i == k:
+            c.x(0)
+            U_gate(i, c)
+            c.x(0)
+        R_gate(i, c)
+    H_gate(q, c)
+    pstar = np.real(np.array(c.expectation([np.array([[1, 1], [1, 1]]) / 2, [0]])))
+    return mod * (2 * pstar - 1)
+
+
+# use original quantum circuit simulate with c
+def simulation():
+    c = tc.Circuit(N, inputs=state)
+    for k in range(len(door)):
+        if door[k][0] == 0:
+            c.rx(door[k][1], theta=ODE_theta[k])
+        if door[k][0] == 1:
+            c.ry(door[k][1], theta=ODE_theta[k])
+        if door[k][0] == 2:
+            c.rz(door[k][1], theta=ODE_theta[k])
+        if door[k][0] == 3:
+            c.rxx(door[k][1], door[k][2], theta=ODE_theta[k])
+        if door[k][0] == 4:
+            c.ryy(door[k][1], door[k][2], theta=ODE_theta[k])
+        if door[k][0] == 5:
+            c.rzz(door[k][1], door[k][2], theta=ODE_theta[k])
+        if door[k][0] == 6:
+            c.crx(door[k][1], door[k][2], theta=ODE_theta[k])
+        if door[k][0] == 7:
+            c.cry(door[k][1], door[k][2], theta=ODE_theta[k])
+        if door[k][0] == 8:
+            c.crz(door[k][1], door[k][2], theta=ODE_theta[k])
+    return c
+
+
+if __name__ == "__main__":
+
+    # l: layers; h and J: coefficient of Hamilton; L_var and L_num: results of variation method and numerical method
+    N = 3
+    l = 2
+    J = 1 / 4
+    dt = 0.05
+    t = 1
+    h = []
+    L_var = []
+    L_num = []
+    x_value = []
+
+    how_variation = 0  # 0 McLachlan 1 time-dependent
+
+    # the priciple correspond with all gates
+    # the first term: 0rx,1ry,2rz,3rxx,4ryy,5rzz,6crx,7cry,8crz;
+    # the second and the third term: num/ctrl+num
+    # f: coefficient with simulation gates in paper
+    door = []
+    h_door = []
+    f = []
+    for k in range(l):
+        for i in range(N):
+            f.append(-0.5j)
+            door.append([0, i])
+        for i in range(N - 1):
+            f.append(-1j)
+            door.append([5, i, i + 1])
+        for i in range(N - 1):
+            f.append(-1j)
+            door.append([3, i, i + 1])
+    for i in range(N):
+        h.append(1)
+        h_door.append([0, i])
+    for i in range(N - 1):
+        h.append(J)
+        h_door.append([5, i, i + 1])
+
+    # initial state
+    state = np.zeros(1 << N)
+    state[0] = 1
+
+    # numerical realize H
+    H = np.zeros((1 << N, 1 << N)) * 1j
```

Comment:
calculate H is this way is slow, please try to use ``tc.quantum.PauliStringSum2COO``